routing/router.go

```diff
@@ -101,11 +102,38 @@ func NewRouter() *mux.Router {
 	return rootRouter
 }
 
+// statusResponseWriter contains a ResponseWriter and a StatusCode to read in the metrics middleware
+type statusResponseWriter struct {
+	http.ResponseWriter
+	StatusCode int
+}
+
+// WriteHeader will be called by any function that needs to set an status code, in this function the StatusCode is also set
+func (srw *statusResponseWriter) WriteHeader(code int) {
+	srw.ResponseWriter.WriteHeader(code)
+	srw.StatusCode = code
+}
+
+// updateMetric evaluates the StatusCode, if there is an error, increase the API failure counter, otherwise save the duration
+func updateMetric(route string, srw *statusResponseWriter, timer *prometheus.Timer) {
+	// Increase the error counter on 500 and 503 errors
+	if srw.StatusCode == http.StatusInternalServerError || srw.StatusCode == http.StatusServiceUnavailable {
+		internalmetrics.GetAPIFailureMetric(route).Inc()
+	}
+	// Always measure the duration even if the API call ended in an error
+	timer.ObserveDuration()
```

Comment:
I know its very little time, but timer.ObserveDuration should be called as soon as we know the API call is done. Now we are timing not only the API call but also the call to get into this function, and the time it takes to invoke that if-stmt and the call to `GetAPIFailureMetric(route).Inc()`. I would just move the timer.ObserveDuration() as the very first thing this method does to get that timer to more accurately reflect the time. Not a big deal, but is more accurate this way. In fact, the most accurate way would be to put this back the way it was - do a `defer timer.ObserveDuration()` as the first defer call in metricHandler. Then you can call this method to check the failures. And really, since its only three lines, I don't think we even need an extra function here - just put that in the metricHandler to make it simpler - but that's just aesthetics - keeping an extra function is fine if it makes things more understandable/readable.
 
```suggestion
	// Always measure the duration even if the API call ended in an error
	timer.ObserveDuration()
	// Increase the error counter on 500 and 503 errors
	if srw.StatusCode == http.StatusInternalServerError || srw.StatusCode == http.StatusServiceUnavailable {
		internalmetrics.GetAPIFailureMetric(route).Inc()
	}
```